cpp/include/cudf/lists/detail/gather.cuh

```diff
@@ -90,12 +92,19 @@ gather_data make_gather_data(cudf::lists_column_view const& source_column,
     count_iter,
     count_iter + offset_count,
     dst_offsets_v.begin<int32_t>(),
-    [gather_map, output_count, src_offsets, src_size] __device__(int32_t index) -> int32_t {
+    [d_source_column = *d_source_column,
+     gather_map,
+     output_count,
+     src_offsets,
+     src_size] __device__(int32_t index) -> int32_t {
       int32_t offset_index = index < output_count ? gather_map[index] : 0;
 
       // if this is an invalid index, this will be a NULL list
       if (NullifyOutOfBounds && ((offset_index < 0) || (offset_index >= src_size))) { return 0; }
 
+      // If the source row is null, the output row size must be 0.
+      if (not d_source_column.is_valid(offset_index)) { return 0; }
```

Comment:
This is probably out of scope for this PR, but my DRY senses are tingling on the logic inside this lambda as it appears to be mostly repeated in a few places.

Reply:
I tried collapsing these lambdas into a separate functor. Because of the differences between them, the functor looked like a contrivance, and not particularly readable.
Unless @nvdbaranec suggests a better approach, I'm tempted to postpone this change.